lib/userializer/composite_serializer.rb

```diff
@@ -0,0 +1,81 @@
+require 'oj'
+
+module USerializer
+  class CompositeSerializer
+    class CompositeObject
+      def initialize(obj, opts = {})
+        @obj = obj
+        @opts = opts
+        @root_key = (
+          opts[:root] ||
+          ActiveSupport::Inflector.underscore(obj.class.name).split('/').last
+        ).to_sym
+        serializer = opts[:serializer]
+        @serializer = if serializer.is_a?(Proc)
+                        serializer
+                      elsif serializer
+                        proc { serializer }
+                      end
+      end
+
+      def merge_root(res, opts)
+        serializer(@obj, opts).merge_root(res, @root_key, true, opts)
+      end
+
+      private
+
+      def serializer(obj, opts)
+        return @serializer.call(obj, opts).new(obj, @opts) if @serializer
+        return obj.serialize if obj.respond_to?(:serialize)
+
+        USerializer.infered_serializer_class(obj.class).new(obj, @opts)
+      end
+    end
+
+    def initialize(objs, opts = {})
+      @opts = opts
+      @objs = compose_objs(objs)
+    end
+
+    def merge_root(res, opts)
+      @objs.each do |obj|
+        obj.merge_root(res, opts)
+      end
+    end
+
+    def to_hash
+      res = {}
+
+      merge_root(res, @opts)
+      res
+    end
+
+    def serialize(*_args)
+      to_hash
+    end
+
+    def to_json(*_args)
+      Oj.dump(to_hash, mode: :compat)
+    end
+
+    private
+
+    def compose_objs(objs)
+      objs.map do |(key, obj)|
+        opts = options_for(key)
+
+        if obj.is_a? Enumerable
+          USerializer::ArraySerializer.new(obj, opts)
+        else
+          CompositeObject.new(obj, opts)
+        end
+      end
+    end
+
+    def options_for(key)
+      @opts.reduce({}) do |acc, (opt_key, values)|
+        acc.merge(opt_key.to_sym => values[key.to_sym])
+      end.compact
```

Comment:
wdyt of

```suggestion
      end.compact.merge(root: key) { |_, x, y| x || y }
```

it would take the key as root for the param?

Reply:
I'm not sure to understand what you want to do ^^